src/mlpack/core/optimizers/saga/saga_impl.hpp

```diff
@@ -0,0 +1,185 @@
+    const UpdatePolicyType& updatePolicy,
+    const DecayPolicyType& decayPolicy,
+    const bool resetPolicy) :
+    stepSize(stepSize),
+    batchSize(batchSize),
+    maxIterations(maxIterations),
+    tolerance(tolerance),
+    shuffle(shuffle),
+    updatePolicy(updatePolicy),
+    decayPolicy(decayPolicy),
+    resetPolicy(resetPolicy)
+{ /* Nothing to do. */ }
+
+//! Optimize the function (minimize).
+template<typename UpdatePolicyType, typename DecayPolicyType>
+template<typename DecomposableFunctionType>
+double SAGAType<UpdatePolicyType, DecayPolicyType>::Optimize(
+    DecomposableFunctionType& function,
+    arma::mat& iterate)
+{
+  // Find the number of functions to use.
+  const size_t numFunctions = function.NumFunctions();
+
+  // To keep track of where we are and how things are going.
+  double overallObjective = 0;
+  double lastObjective = DBL_MAX;
+
+  // Initialize the update policy.
+  if (resetPolicy)
+    updatePolicy.Initialize(iterate.n_rows, iterate.n_cols);
+
+  // Now iterate!
+  arma::mat gradient(iterate.n_rows, iterate.n_cols);
+  arma::mat gradient0(iterate.n_rows, iterate.n_cols);
+  arma::mat iterate0;
+
+  // Find the number of batches.
+  size_t numBatches = numFunctions / batchSize;
+  if (numFunctions % batchSize != 0)
+    ++numBatches; // Capture last few.
+
+  const size_t actualMaxIterations = (maxIterations == 0) ?
+               std::numeric_limits<size_t>::max() : maxIterations;
+  for (size_t i = 0; i < actualMaxIterations; ++i)
+  {
+    // Calculate the objective function.
+    overallObjective = 0;
+    for (size_t f = 0; f < numFunctions; f += batchSize)
+    {
+      const size_t effectiveBatchSize = std::min(batchSize, numFunctions - f);
+      overallObjective += function.Evaluate(iterate, f, effectiveBatchSize);
+    }
+
+    if (std::isnan(overallObjective) || std::isinf(overallObjective))
+    {
+      Log::Warn << "SAGA: converged to " << overallObjective
+                << "; terminating  with failure.  Try a smaller step size?"
+                << std::endl;
+      return overallObjective;
+    }
+
+    if (std::abs(lastObjective - overallObjective) < tolerance)
+    {
+      Log::Info << "SAGA: minimized within tolerance " << tolerance
+                << "; terminating optimization." << std::endl;
+      return overallObjective;
+    }
+
+    lastObjective = overallObjective;
+
+    // initialize the gradients.
+    size_t effectiveBatchSize = std::min(batchSize, numFunctions);
+    arma::cube tableOfGradients(iterate.n_rows, iterate.n_cols, numBatches);
+    arma::mat avgGradient(iterate.n_rows, iterate.n_cols);
+
+    for (size_t f = 0, b=0; f < numFunctions;
+      /* incrementing done manually */)
+    {
+      // Find the effective batch size (the last batch may be smaller).
+      effectiveBatchSize = std::min(batchSize, numFunctions - f);
+
+      function.Gradient(iterate, f, gradient, effectiveBatchSize);
+      tableOfGradients.slice(b) = gradient;
```

Comment:
We could avoid the extra copy here, if we use ``tableOfGradients.slice(b) `` inside the ``Gradient`` function instead of the ``gradient`` parameter.